matlab/src/feather_writer.cc

```diff
@@ -0,0 +1,365 @@
+    case mxINT16_CLASS:
+      return WriteNumericData<arrow::Int16Type>(data, validity_bitmap);
+    case mxINT32_CLASS:
+      return WriteNumericData<arrow::Int32Type>(data, validity_bitmap);
+    case mxINT64_CLASS:
+      return WriteNumericData<arrow::Int64Type>(data, validity_bitmap);
+
+    default: {
+      mexErrMsgIdAndTxt("MATLAB:arrow:UnsupportedArrowType",
+                        "Unsupported arrow::Type '%s' for variable '%s'",
+                        mxGetClassName(data), type.c_str());
+    }
+  }
+
+  // We shouldn't ever reach this branch, but if we do, return a null unique_ptr.
+  return std::unique_ptr<arrow::Array>();
+}
+
+}  // namespace internal
+
+arrow::Status FeatherWriter::Open(const std::string& filename,
+                                  std::shared_ptr<FeatherWriter>* feather_writer) {
+  // Allocate shared_ptr out parameter.
+  *feather_writer = std::shared_ptr<FeatherWriter>(new FeatherWriter());
+
+  // Open a FileOutputStream corresponding to the provided filename.
+  std::shared_ptr<arrow::io::OutputStream> writable_file(nullptr);
+  ARROW_RETURN_NOT_OK(arrow::io::FileOutputStream::Open(filename, &writable_file));
+
+  // TableWriter::Open expects a shared_ptr to an OutputStream.
+  // Open the Feather file for writing with a TableWriter.
+  ARROW_RETURN_NOT_OK(arrow::ipc::feather::TableWriter::Open(
+      writable_file, &(*feather_writer)->table_writer_));
+
+  return arrow::Status::OK();
+}
+
+// Write table metadata to the Feather file from a mxArray*.
+arrow::Status FeatherWriter::WriteMetadata(const mxArray* metadata) {
+  // Verify that all required fieldnames are provided.
+  internal::ValidateMxStructField(metadata, "Description", mxCHAR_CLASS, true);
+  internal::ValidateMxStructField(metadata, "Version", mxDOUBLE_CLASS, false);
+  internal::ValidateMxStructField(metadata, "NumRows", mxDOUBLE_CLASS, false);
+  internal::ValidateMxStructField(metadata, "NumVariables", mxDOUBLE_CLASS, false);
+
+  // Convert Description to a std::string and set on FeatherWriter and TableWriter.
+  std::string description =
+      internal::MxArrayToString(mxGetField(metadata, 0, "Description"));
+  this->description_ = description;
+  this->table_writer_->SetDescription(description);
+
+  // Get the NumRows field in the struct array and set on TableWriter.
+  this->num_rows_ = static_cast<int64_t>(mxGetScalar(mxGetField(metadata, 0, "NumRows")));
+  this->table_writer_->SetNumRows(this->num_rows_);
+
+  // Get the total number of variables. This is checked later for consistency with
+  // the provided number of columns before finishing the file write.
+  this->num_variables_ =
+      static_cast<int64_t>(mxGetScalar(mxGetField(metadata, 0, "NumVariables")));
+
+  // Store the version information so we could handle possible future API changes.
+  this->version_ = static_cast<int32_t>(mxGetScalar(mxGetField(metadata, 0, "Version")));
+
+  return arrow::Status::OK();
```

Comment:
It doesn't seem like this function is capable of returning an error Status

Reply:
Fixed.